### PR TITLE
Improve lexer error recovery with interpolated strings.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
The current lexer strategy with interpolations is to bail early when it encounters something like a spurious newline in a non-verbatim interpolation.  This means that simple user cases like:

```c#
var v = $"Hello { @""
                   World
                   "" } !";
```

Go entirely off the rails.  since the lexer stops processing that at: `var v = $"Hello { @""` leaving the rest to be restarted.  This of course throws everything off as the contents of the string become code and then later quotes look to start new strings (like `"; ...`).  Furthermore the close curlies found later then just destroy the rest of hte parsing stack, leading to everything being awful.

This PR changes our strategy here to actually recognize when we are running into errant newlines (either in the interpolation code, or comments, or multiline verbatim literals), and while it treats that as an error, it still allows normal lexing of these constructs to proceed, producing a good tree with a good error message.

This cleanup helps the Raw-String-Literals work (esp. around how it handles interpolations).  However, it can be done independently to keep things tidier and simpler.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:152]: this file will be easier to view with whitspace changes off.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:160]: no need for this 'if' check, the callers ensure it.  so this became a Debug assert, with no else.

**Comment by jcouv** [comment]
Is this related to the raw string literal work? If so, should this go into the feature branch so that you don't have to wait for another merge from `main`?

**Comment by CyrusNajmabadi** [comment]
> Is this related to the raw string literal work? 

Tangentially.  THis is pure cleanup and improvement though that can go in outside of the raw string literal work.  that's why i decoupled it and brought it into main.

File: src/EditorFeatures/CSharpTest/Diagnostics/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProviderTests.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProviderTests.cs:99]: these results are actually preferred as we do not have the bogus space after the WhenFalse part.

File: src/EditorFeatures/CSharpTest/Diagnostics/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProviderTests.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProviderTests.cs:91]: preferred as this is symettrical with what we do with Leading1.  both the open and close parens tightly bind the expression code itself, not the surrounding trivia.

**Comment by CyrusNajmabadi** [comment]
tagging @jcouv @333fred or anyone on the compiler who would like to take a look :)

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:174]: >goign [](http://example.com/codeflow?start=95&length=5)

Typo

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:1985]: >9000 [](http://example.com/codeflow?start=79&length=4)

Needs to be a real number

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:298]: >' [](http://example.com/codeflow?start=117&length=1)

No '

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:548]: Braces for consistency with the rest of the function please.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:641]: Same comment about braces

Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:641]: >!= null [](http://example.com/codeflow?start=46&length=7)

Could be `errorCode is ErrorCode value` to avoid the double null-check in Value.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:731]: >an [](http://example.com/codeflow?start=42&length=2)

an -> a

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:735]: Braces

File: src/Compilers/CSharp/Portable/CSharpResources.resx
Comment (333fred) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:6798]: Nit: I think it would read better as `Newlines are not` rather than singular.

**Comment by 333fred** [comment]
Done review pass (commit 14). Nothing major.

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:1985]: 9000 is an integer. Which is one of the real numbers

Comment (333fred) [reply src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:1985]: Sure, but it's not the next number in sequence here. It jumps a bunch of IDs.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:541]: made braces consistent.

File: src/Compilers/CSharp/Portable/CSharpResources.resx
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:6798]: sgtm

**Comment by CyrusNajmabadi** [comment]
@333fred all feedback responded to.  additional cleanup applied.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:5]: this file will be easier to view with whitspace changes off.

**Review by 333fred — APPROVED** [review]
LGTM (commit 30)

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:1984]: Please consider using the existing naming style - something like `ERR_NewlinesNotAllowedInNonVerbatimInterpolatedString`.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:1984]: sure.  though i personally prefer the form i used here (also used uniformly in the ide codebase) because you can tell the casing of stuff.  though that matters more for us as we have a lot of places where we need to use Title Case  for headers and whatnot.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: ok.  renamed.

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:1984]: FYI, there will shortly be a merge conflict with PR [https://github.com/dotnet/roslyn/pull/55042](https://github.com/dotnet/roslyn/pull/55042)
The merging of a C# 10 feature will take precedence. So I'll let you deal with the conflict. Sorry about that

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:39]: While I understand that there are some exceptions in the compiler codebase, our guideline is still to keep braces on `if` statements. This file seems to have been consistent in using them, so we should stick with that. Also, it's not necessary to this PR.
Same applies to other instances below. Thanks

nit: I'd keep the explicit type as well, since `char` is barely longer than `var` anyways.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:225]: Have you considered only holding onto only one error (like before), but adding a flag to signify we've encountered an unrecoverable error?
If we did that, I suspect the earliest error would naturally win (as it would be produced first).

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:552]: nit: consider naming discarded argument

File: src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs
Comment (jcouv) [inline src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:93]: I didn't follow the example. We need some parenthesized expression containing a conditional to get here

File: src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs
Comment (jcouv) [inline src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:81]: Do we need to add tests to cover this change?

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs:451]: >CS9000ERR [](http://example.com/codeflow?start=20&length=9)

Let's do a bulk replace on `CS9000` and `CS9001` with final error code before merging

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs:1266]: >CS8077ERR [](http://example.com/codeflow?start=20&length=9)

nit: no error in this scenario

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs:451]: Consider also doing a bulk replace on `Newlines_are_not_allowed_inside_a_non_verbatim_interpolated_string`

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 32)

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:1984]: i originally started with a gap, since this woudl be a c#11 feature.  i was asked to get rid of that gap :)  lmk what i should do here.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:39]: > While I understand that there are some exceptions in the compiler codebase, our guideline is still to keep braces on if statements

Sure.  I can add these back in.  However, my general feedback is that hte compiler codebase tends to be pretty 'all over the map'.  It would be great if some consistency passes could be done to do it so that code would be definitively in-line or out-of-line with the patterns of the rest.  :)

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:231]: yeah, i actually started with that pattern, but i did find it a little more confusing in terms of how every place that updated errors had to manage checking that and processing accordingly.  i actually thought this was really clean in terms of what was going on, with only the additional wrinkle of checking the starting position.

but i may not toally be getting your idea here.  i'd be happy to collaborate on this :)

File: src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:92]: updated the comment to explain this.

File: src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:81]: there are many :)

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalErrorTests.cs:451]: yup yup.

File: src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs
Comment (jcouv) [reply src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:81]: I saw tests where formatting was adjusted, but not new tests for situation with newline. That's why I was asking

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:81]: ah gotcha.  we don't have new tests there.  THey're already covered by existing tests.  I'll find the names for you of examples of this.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:231]: Just to clarify my idea (maybe it doesn't work that well, not sure):
1. `InterpolatedStringScanner` has `public SyntaxDiagnosticInfo error;` and `public bool encounteredUnrecoverableError = false;`
2. All places that do `UnrecoverableError = ...` or `UnrecoverableError ??= ...` use a helper method (`SetUnrecoverableError`) that sets `error` if it's the first one and `encounteredUnrecoverableError = true`.
3. The one place that does `RecoverableError ??= ...` can just do `error ??= ...`
4. `RecoveringFromRunawayLexing` checks `encounteredUnrecoverableError`
5. We should be able to remove the handling in `ScanInterpolatedStringLiteralTop` for picking which error to report

File: src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:81]: TestAddParenthesesWhileTyping1-5 and TestAddParenthesesWithCS1026PresentBeforeFixIsApplied1-3

Thanks!

Comment (jcouv) [reply src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:81]: That's fine. No need, just knowing we have them is enough :-)

File: src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs
Comment (jcouv) [reply src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:92]: Thanks!

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:1984]: This is a bugfix level change. It will ship in dev17, since the PR is targeting the `main` branch.
So it should be fine to use the next available error code (after I merge parameterless struct constructors). No need to use a new range.

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 38)

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:231]: ok.  i'll try that!

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs]: Ok.  Done in https://github.com/dotnet/roslyn/pull/54875/commits/59e158d6cbd4b92917e4f433d4c82b36e798dbef.  Pro: it works.  Con: it technically means we may allocate more in error cases as we now always make the error prior to checking if we already have an error.

I could update this to pass in arguments, but there are some params cases, and we'll have to allocate.  

THat said, it's likely not a big deal as this is only for error conditions and it's unlikely that this would matter in practice.  Up to you if we should keep this or rollback :)

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs]: I think it's a positive change. I'm not worried about this allocation, but passing in arguments would seem sensible if it helps. Worst case it's possible to add a null check at callsite.
Thanks!

**Comment by jcouv** [comment]
Looks like a couple of tests were affected (break in CI)

**Comment by CyrusNajmabadi** [comment]
> Looks like a couple of tests were affected (break in CI)

Yeah, looks like the approach of picking the earliest error was necessary.  I've moved to your model, but have preserved that logic.

**Review by 333fred — APPROVED** [review]
LGTM (commit 42)

**Review by jcouv — APPROVED** [review]
LGTM (iteration 43). Don't forget to squash please. Thanks

**Comment by CyrusNajmabadi** [comment]
Yup.  I'm set to squash things.  Thanks!

**Comment by 333fred** [comment]
@CyrusNajmabadi looks like there's a legitimate test failure.

**Pull request merged**
